core/utils.py

```diff
@@ -41,205 +46,40 @@ class TibannaStartException(Exception):
-                               Prefix=prefix)
-
-    def s3_delete_dir(self, prefix):
-        # one query get list of all the files we want to delete
-        obj_list = s3.list_objects(Bucket=self.outfile_bucket,
-                                   Prefix=prefix)
-        files = obj_list.get('Contents', [])
-
-        # morph file list into format that boto3 wants
-        delete_keys = {'Objects': []}
-        delete_keys['Objects'] = [{'Key': k} for k in
-                                  [obj['Key'] for obj in files]]
-
-        # second query deletes all the files, NOTE: Max 1000 files
-        if delete_keys['Objects']:
-            s3.delete_objects(Bucket=self.outfile_bucket,
-                              Delete=delete_keys)
-
-    def read_s3_zipfile(self, s3key, files_to_extract):
-        s3_stream = self.read_s3(s3key)
-        bytestream = BytesIO(s3_stream)
-        zipstream = ZipFile(bytestream, 'r')
-        ret_files = {}
-
-        for name in files_to_extract:
-            # search subdirectories for file with name
-            # so I don't have to worry about figuring out the subdirs
-            zipped_filename = find_file(name, zipstream)
-            if zipped_filename:
-                ret_files[name] = zipstream.open(zipped_filename).read()
-        return ret_files
-
-    def unzip_s3_to_s3(self, zipped_s3key, dest_dir, retfile_names=None, acl=None):
-        if retfile_names is None:
-            retfile_names = []
-
-        if not dest_dir.endswith('/'):
-            dest_dir += '/'
-
-        s3_stream = self.read_s3(zipped_s3key)
-        # read this badboy to memory, don't go to disk
-        bytestream = BytesIO(s3_stream)
-        zipstream = ZipFile(bytestream, 'r')
-
-        # directory should be first name in the list
-        file_list = zipstream.namelist()
-        basedir_name = file_list.pop(0)
-        assert basedir_name.endswith('/')
-
-        ret_files = {}
-        for file_name in file_list:
-            # don't copy dirs just files
-            if not file_name.endswith('/'):
-                s3_file_name = file_name.replace(basedir_name, dest_dir)
-                s3_key = "https://s3.amazonaws.com/%s/%s" % (self.outfile_bucket, s3_file_name)
-                # just perf optimization so we don't have to copy
-                # files twice that we want to further interogate
-                the_file = zipstream.open(file_name, 'r').read()
-                file_to_find = file_name.split('/')[-1]
-                if file_to_find in retfile_names:
-                    ret_files[file_to_find] = {'s3key': s3_key,
-                                               'data': the_file}
-                self.s3_put(the_file, s3_file_name, acl=acl)
-
-        return ret_files
-
-
-def find_file(name, zipstream):
-    for zipped_filename in zipstream.namelist():
-        if zipped_filename.endswith(name):
-            return zipped_filename
+def get_source_experiment(input_file_uuid, ff_keys, ff_env):
```

Comment:
is it possible to convert this to a single query?  Do we need to?  On average how many times are we running this loop?

Reply:
Yeah, this could be done with one frame=embedded query to `inf_uuid` (instead of multiple queries to a file and all of its experiments). This was the original way it was written and I didn't change it ... want me to?